src/main/scala/tech/cryptonomic/conseil/Lorre.scala

```diff
@@ -219,7 +202,32 @@ object Lorre extends App with TezosErrors with LazyLogging with LorreAppConfig w
     def processBakingAndEndorsingRights(fetchingResults: tezosNodeOperator.BlockFetchingResults): Future[Unit] = {
       import cats.implicits._
       val bh = fetchingResults.map(_._1.data.hash)
-      (tezosNodeOperator.getBatchBakingRights(bh), tezosNodeOperator.getBatchEndorsingRights(bh)).mapN {
+
+      def safeHead[A](list : List[A]): Option[A] = {
+        list match {
+          case Nil => None
+          case x :: _ => Some(x)
+        }
+      }
+
+      val cycle = safeHead(fetchingResults).flatMap{
+        _._1.data.metadata match {
+          case GenesisMetadata => None
+          case BlockHeaderMetadata(_,_,_,_,_,level) => Some(level.cycle)
+        }
+      }
+
+      val governancePeriod = safeHead(fetchingResults).flatMap{
+        _._1.data.metadata match {
+          case GenesisMetadata => None
+          case BlockHeaderMetadata(_,_,_,_,_,level) => Some(level.voting_period)
+        }
+      }
+
+      val blockHashesWithCycleAndGovernancePeriod = bh.map(hash => (cycle, governancePeriod, hash))
+
+      (tezosNodeOperator.getBatchBakingRights(blockHashesWithCycleAndGovernancePeriod),
+        tezosNodeOperator.getBatchEndorsingRights(blockHashesWithCycleAndGovernancePeriod)).mapN {
```

Comment:
Here it looks like we have to adapt the fetcher to accept the triple, but still use only the block-hash internally.

Why not fetch the rights passing only the block-hash, and then "build" the triplet only when calling the `writeXXXRights` methods?

We can, e.g., build an index to map hashes to the correct cycle/period
```scala
//define two extractor functions
val readCycle: BlockData => Option[Int] = data => ...
val readPeriod: BlockData => Option[Int] = data => ...

(getBatchBakingRights..., getBatchEndorsingRights...).mapN {
  case (br, er) =>
    (db.run(TzDB.writeBakingRights(withLevelData(br)), db.run(TzDB.writeEndorsingRights(withLevelData(er)))
      .mapN((_, _) => ())
}

//define an index and a way to attach cycle/period to rights map
//the index associated with the hash the actual level data
val levelDataIndex: Map[BlockHash, (Option[Int], Option[Int])] = fetchingResults.map {
  case (Block(data, _, _), _) =>
    data.hash -> (readCycle(data), readPeriod(data))
}.toMap

//this should convert inputs to add the missing level info, by mapping the keys and pairing them with those from the index
def withLevelData[T](tMap: Map[BlockHash, T]): Map[(Option[Int], Option[Int], BlockHash), T] = ...
```

Reply:
This would by-pass the need to update the rights fetchers completely.

Reply:
I agree with this fundamentally, but may leave for another pr, as we have to get this functionality out soon.

Reply:
Will refactor to abide by this soon.